### PR TITLE
3391 crypto cleanup

### LINKED ISSUE #3391: Crypto: drop support for Botan and OpenSSL variants
The crypto plugin(s) consist(s) in variants with the following crypto libraries in use:

- libgcrypt,
- OpenSSL, and
- Botan.

The variants have been introduced for academic research, which was finished in 2018. There is no other benefit of supporting several crypto libraries.

On the downside the maintenance effort of supporting the libraries mentioned above is increasing (e.g. #3385 ).

I think **libgcrypt** is the crypto library that best fits our needs (and has the nicest code-base). Also I do not expect the crypto library to cause increased maintenance effort.

Therefore I propose to drop the support of:

- OpenSSL, and
- Botan

and remove their corresponding compilation variants from the crypto plugin.

### PR DESCRIPTION
**Pull request by petermax2**
## Basics

Closes #3391 .

These points need to be fulfilled for every PR:

- [x] Short descriptions of your changes are in the release notes
      (added as entry in `doc/news/_preparation_next_release.md` which
      contains `_(my name)_`)
      **Please always add something to the release notes.**
- [x] Details of what you changed are in commit messages
      (first line should have `module: short statement` syntax)
- [ ] References to issues, e.g. `close #X`, are in the commit messages.
- [ ] The buildservers are happy.
- [ ] The PR is rebased with current master.

If you have any troubles fulfilling these criteria, please write
about the trouble as comment in the PR. We will help you.
But we cannot accept PRs that do not fulfill the basics.

## Checklist

Check relevant points but **please do not remove entries**.
For docu fixes, spell checking, and similar none of these points below
need to be checked.

- [ ] I added unit tests for my code
- [x] I fully described what my PR does in the documentation
      (not in the PR description)
- [x] I fixed all affected documentation
- [ ] I added code comments, logging, and assertions as appropriate (see [Coding Guidelines](https://master.libelektra.org/doc/CODING.md))
- [ ] I updated all meta data (e.g. README.md of plugins and [METADATA.ini](https://master.libelektra.org/doc/METADATA.ini))
- [ ] I mentioned every code not directly written by me in [THIRD-PARTY-LICENSES](doc/THIRD-PARTY-LICENSES)

## Review

Reviewers will usually check the following:

- [ ] Documentation is introductory, concise, good to read and describes everything what the PR does
- [ ] Examples are well chosen and understandable
- [ ] Code is conforming to [our Coding Guidelines](https://master.libelektra.org/doc/CODING.md)
- [ ] APIs are conforming to [our Design Guidelines](https://master.libelektra.org/doc/DESIGN.md)
- [ ] Code is consistent to [our Design Decisions](https://master.libelektra.org/doc/decisions)

## Labels

If you are already Elektra developer:

- Add the "work in progress" label if you do not want the PR to be reviewed yet.
- Add the "ready to merge" label **if the basics are fulfilled** and you also
  say that everything is ready to be merged.

### CODE REVIEW

**Comment by mpranj** [comment]
Thank you so much for working on this!

Don't get distracted by:
```
dh_install: libelektra4-crypto missing files: usr/lib/*/elektra4/libelektra-crypto_gcrypt.
```

I'll remove the corresponding lines from the debian branch. I think after that there should be no other build failures.

File: src/plugins/crypto/README.md [hunk 1]
```diff
@@ -21,33 +21,12 @@ It also needs to be decrypted whenever an admissible access (read) is being perf
...
-
-#ifdef ELEKTRA_CRYPTO_API_BOTAN
-
-- `libbotan1.10-dev` or `botan-devel`
-
-#endif
-
 ### GnuPG (GPG)
 
 GPG is a run-time dependency for all crypto plugin variants.
```
Comment (petermax2) [inline src/plugins/crypto/README.md:32]: TODO: replace "all crypto plugin variants" with "of the crypto plugin".

**Comment by petermax2** [comment]
We have a green build :tada: thanks to @mpranj .

Please have a closer look at my changes, maybe I forgot to update some docs or build server stuff?

**Comment by petermax2** [comment]
> Looks good as far as I can tell. I noticed some additional stuff we can remove, which I committed in my version of the repository. Is it okay for you, if I just add these changes to your PR?

Sure, go ahead please.

**Comment by sanssecours** [comment]
> Sure, go ahead please.

Thank you. Hopefully my changes do not break anything.

**Comment by mpranj** [comment]
I found basically the same remainders which are fixed by @sanssecours. Do we leave `./src/plugins/crypto/compile_openssl.c` because it is used in our CMake `./scripts/cmake/Modules/FindLibOpenSSL.cmake`?

**Comment by petermax2** [comment]
> Do we leave ./src/plugins/crypto/compile_openssl.c because it is used in our CMake ./scripts/cmake/Modules/FindLibOpenSSL.cmake?

I think the `curl` plugin still depends on it. I got compile errors after removing `compile_openssl.c`.

**Comment by mpranj** [comment]
There are multiple other openssl dependencies. Let's leave it in.

There is one more advanced variable which can probably be removed: `./scripts/cmake/ElektraCache.cmake:392:	BOTAN_INCLUDE_DIRS`

File: src/plugins/crypto/testmod_crypto.c [hunk 2]
```diff
@@ -1,12 +1,12 @@
 /**
  * @file
  *
- * @brief test suite for the crypto plugin.
- * Contains shared functions for all compile variants.
+ * @brief test suite for the crypto plugin (gcrypt compile variant)
```
Comment (mpranj) [inline src/plugins/crypto/testmod_crypto.c:4]: There should be only one variant left ? (i.e. there are no compile variants)

**Comment by petermax2** [comment]
`testmod_crypto` does not shut down `gpg-agent` properly on Debian unstable. I will investigate.

**Comment by mpranj** [comment]
If it helps, you can get the docker image by replacing hub.libelektra.org with hub-public.libelektra.org like:
```
docker pull hub-public.libelektra.org/build-elektra-debian-sid:202004-28c863836c6dc56d8e7e0f9c4b87fdfda27f934b58885224f9a7fc61a5c00585
```

**Comment by markus2330** [comment]
Thank you all so much for cleaning up so properly!

**Comment by petermax2** [comment]
As usually with `gpg-agent` related topics I can not reproduce the issue locally, even when using the original Docker container :confused: Let's try to use `gnupg2` and hope that the problem is related to GnuPG version 1.

**Comment by mpranj** [comment]
> Let's try to use gnupg2 and hope

I fear that gnupg is just a synonym for gnupg2 on debian unstable, but let's see anyway.

**Comment by petermax2** [comment]
jenkins build please

**Comment by mpranj** [comment]
@petermax2 sorry, seems that there was a leftover botan dependency in the debian branch. Should be fixed with 05a002af33ad7143a8c6348f8c63d02ad76fb17b. I triggered a rebuild.

**Comment by petermax2** [comment]
It's good to have a second round of building on Debian unstable. Maybe the gpg-agent error does not occur regularely.

**Comment by petermax2** [comment]
The `gpg-agent` related error did not occur again, but we should keep an eye on it.

Ready to merge? What do you think?

**Comment by petermax2** [comment]
We actually observed the same problem as described in #2341 !

**Comment by mpranj** [comment]
> We actually observed the same problem as described in #2341 !

Looks like this also happened on [Fedora 31](https://build.libelektra.org/job/libelektra/job/PR-3395/13/artifact/artifacts/fedora-31-full/DynamicAnalysis.xml). Let's keep an eye on that too if something can be done.

This is definitely ready though.

**Comment by mpranj** [comment]
Thank you so much for fixing the problems and cleaning up here! Great to see some improvements on the crypto plugin.

**Pull request merged**